File: random-forest.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed Apr  9 13:30:52 2025

@author: marybethwalsh
"""

# Random Forest Regression

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Student_Performance.csv')
X = dataset.iloc[:, :-1].to_numpy()
y = dataset.iloc[:, -1].to_numpy()

# Encoding categorical data
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [2])],
                       remainder='passthrough')
X = ct.fit_transform(X)

# Avoiding the Dummy Variable Trap
X = X[:, 1:]
from sklearn.model_selection import train_test_split
X_train_raw, X_test_raw, Y_train_raw, Y_test_raw = train_test_split(X, y, test_size=0.25)


# Training the Random Forest Regression model on the whole dataset
from sklearn.ensemble import RandomForestRegressor
regressor = RandomForestRegressor(n_estimators = 500)  
regressor.fit(X_train_raw, Y_train_raw)



y_pred_rf=regressor.predict(X_test_raw)

# Evaluating the Model Performance (Preferred)
## Universal for all regression models
## Most convincing when used on test sets
## Applicable to both training and test sets
from sklearn.metrics import r2_score
r2 = r2_score(Y_test_raw, y_pred_rf)
print(r2)
r2_adjusted = 1 - (1 - r2) * (len(Y_test_raw) - 1) / (len(Y_test_raw) - 5 - 1)
print('adjusted ', r2_adjusted)


# Visualizing the Random Forest Regression results (for higher resolution and
X_past_score=X_test_raw[:,2]
plt.scatter(X_past_score, Y_test_raw,color = 'red')
plt.scatter(X_test_raw[:, 2], y_pred_rf, color='blue', marker='x', label='Predicted')

plt.title('Random Forest Regression')
plt.xlabel('Past Test Scores')
plt.ylabel('Current Performance Index')
plt.show()

File: random-forest.R
# Random Forest Regression

# Importing the dataset
dataset = read.csv('Student_Performance.csv')

# Encoding categorical data (extra curriculars)
dataset$Extracurricular.Activities = factor(dataset$Extracurricular.Activities,
                                            levels = c('No', 'Yes'),
                                            labels = c(0,1))

# Fitting Random Forest Regression to the whole dataset
library(e1071)
library(caTools)
split <- sample.split(dataset$  Performance.Index, SplitRatio = 0.75)
training_set <- subset(dataset, split == TRUE)
test_set <- subset(dataset, split == FALSE)
library(randomForest)

regressor = randomForest(formula =  Performance.Index ~.,
                         data = training_set,
                         ntree = 500)

# Making a prediction
y_pred <-predict(regressor,newdata=test_set)
print(y_pred)

ssr = sum((test_set$Performance.Index - y_pred) ^ 2)
sst = sum((test_set$Performance.Index - mean(test_set$Performance.Index)) ^ 2)
r2 = 1 - (ssr/sst)
print(r2)
r2_adjusted = 1 - (1 - r2) * (length(test_set$Performance.Index) - 1) / (length(test_set$Performance.Index) - 6 - 1)
print(paste('adjusted rf', r2_adjusted))

# Visualizing the Random Forest Regression results
library(ggplot2)
x_range <- range(dataset$Previous.Scores)
y_range <- range(dataset$Performance.Index)
print(ggplot() +
        geom_point(aes(x = test_set$Previous.Scores, y = test_set$Performance.Index), colour = 'red') +
        geom_point(aes(x = test_set$Previous.Scores, y = predict(regressor, newdata = test_set)), colour = 'blue') +
        ggtitle('Previous score vs Current Performance Index ') +
        xlab('Previous Score') + ylab('Current Performance Index') +
        xlim(x_range) + ylim(y_range))    